### ISSUE TITLE
Linear Classifier

### ISSUE THREAD

**Issue by cleebp** [cleebp/csc-522-team5#5]
Here is one using SVM: http://www.ijceronline.com/papers/Vol2_issue5/N02512541257.pdf

**Comment by cleebp** [comment]
https://www.kaggle.com/wezemer/data-science-bowl-2017/selective-enhancement-filters

`ln[16]` and on might be relevant for feature extraction, I followed this code loosely for ROI so we should at least be able to use the `maxima_divergence` function if needed.

**Comment by mohits19** [comment]
Rishabh has pushed code for feature extraction and classification. 
classifierC.py uses Neural Network. The code is complete but takes a long time to run. We haven't been able to do that. Brian, could you try to run it on your laptop
classifier.py uses the code in tutorial. This gives a lot of errors which we havent been able to figure out yet. I am gonna work on that today.

**Comment by cleebp** [comment]
Yeah I will try running it on my lab computer, could you give me some info on how to run it all and I will report back with what it spits out (I have all the stage1 files on this machine too).

**Comment by Rishabh2693** [comment]
First you need to install tensorflow.. then run new.py to get the.npy files.. and then run classifierC.py.. You must also have 'sample_images/' as input dir, and following as o/p dir within your git repo..
and before running classifier, You need to split and put the npy files generated in prepd into these directories 'prepd_stage1/train/'
'prepd_stage1/valid/'
'prepd_stage1/test/')

**Comment by mohits19** [comment]
I updated the code a little bit.
Now you can run "new.py" for the preprocessing + ROI.
Then run "data_dir.py" to split .npy files into the three folders.
Then run "classifierC.py"

I guess this method does NOT need us to separately extract features since it is extracting shape from image and training model based on that.

We should try to run this code first on sample_images to make sure it is working. may not get good accuracy here since test and valid sets are too small.
Then run on stage1 labels. No idea how long this will take!! Hopefully we get a good accuracy here!!

**Comment by mohits19** [comment]
In case classifierC.py does not work, I am working on classifier.py as well. The problem with current code is directly classifying without training a model first. The code for this is at https://github.com/rdcolema/dsb/blob/master/resnet_xgb.ipynb and https://github.com/booz-allen-hamilton/DSB3Tutorial/blob/master/tutorial_code/LUNA_train_unet.py.

I will continue working on this tomorrow morning. Meanwhile, somebody make sure to run classifierC.py as mentioned above and let us know the result.

**Comment by mohits19** [comment]
I also pushed train_xgboost.py from https://github.com/rdcolema/dsb/blob/master/resnet_xgb.ipynb
Still giving errors but I will start debugging that later today.

**Comment by cleebp** [comment]
Quick question, I just installed all the required CUDA components neccesary to run this on my GPU to speed things up, I was now looking at TensorFlow and it looks like it only supports pip installation for Python3.5.x? If this is the case I'll need to reconfigure my system and dependencies for that since we have been using Python2.7 up till now, did you find a different way to install it for Python2 or did you switch to Python3?

**Comment by Rishabh2693** [comment]
I was using an older version of tensorflow that was run using python2.. that had been pre installed on a back, for an independent study. It wasn't using gpu..

**Comment by cleebp** [comment]
hmmm okay, it looks like I might have to use the newest tensorflow which only supports Python3 if I want to use GPU support, I might try to figure that out since running this on my GPU will save us decades of running time on the /stage1 files... 😆

**Comment by cleebp** [comment]
Been bashing my head against installing depencies all morning, I finally got tensorflow working correctly with CUDA and cudnn on my GPU, but then I realized that to run our code you really need Anaconda for the BLAS/LAPACK packaged in - however it is surprisingly difficult to install tensorflow with GPU support on Anaconda. Have to drive back to Raleigh now but I'll try to get this all figured out tonight so I can start running the code.

**Comment by mohits19** [comment]
Not able to train using xgboost. Seems like simple code that should work!
Getting all kinds of errors that i am unable to debug.

**Comment by mohits19** [comment]
Finally made some progress!! :) used code from https://www.kaggle.com/sentdex/data-science-bowl-2017/first-pass-through-data-w-3d-convnet/notebook.
I pushed the code on a separate branch called "conv-nn"

Run "preprocess.py" to produce "muchdata-50-50-20.npy"
Then run "conv-nn.py" to train the neural network.

![capture](https://cloud.githubusercontent.com/assets/21266662/25080964/71240a1c-2315-11e7-9189-1f3e494a94d9.PNG)
Still only on 5/10 epochs for sample images but it seems to work (I will wait to see it till the end).
Definitely faster than classifierC.py was but still slow. I am running it on PuTTy, so you guys can probably do it faster.

Tomorrow morning, we can run on all stage1 images!!
Still need to figure out how to test the model though!!

**Comment by cleebp** [comment]
awesome I'll try to get it all running on my system today, I should be able to do all the TensorFlow calculations on my 1070 which will speed things up 😺

**Comment by mohits19** [comment]
Awesome!! Right now I am downloading stage1 files on VCL. Then I will try to run it on that.

**Comment by cleebp** [comment]
Just set everything up on my machine, running it with my GPU it takes like 1 minute for the sample images and I get the following output:

![cnn](https://cloud.githubusercontent.com/assets/3922117/25092386/6c75d3b0-235b-11e7-980d-d34308f2cb89.png)

If this looks about right I'll go ahead and start running on stage1/, anything else we want to add to the output before running it on the data?

I'm adding timing prints right now for my own purposes, it looks like each epoch is taking ~.5 seconds which is awesome (assuming its working correctly)

**Comment by mohits19** [comment]
That looks fine!
the validation set is only of length 2 right now, that why accuracy is 0,1 or 0.5
When running on stage1 change it to 100.

**Comment by cleebp** [comment]
Sounds good, running pre-process on stage1/ right now, this is gonna take awhile...

**Comment by cleebp** [comment]
So I ran preprocessing and got my muchdata on the stage1 files, it took roughly 20 minutes.

Just ran the actual CNN and each epoch took about 30 seconds, I received a weird error each time though that looked like this: `Invalid argument: Input to reshape is a tensor with 42500 values, but the requested shape requires a multiple of 50000`. It didnt seem to stop everything from working though but it is weird, someone posted on the notebook with the same error when running on GPU version.

Regardless this is what my output looks like:

`Done. Finishing accuracy:`
`Accuracy: 0.69`
`fitment percent: 0.9992289899768697`
`Total running time: 320.111608505249 seconds.`

So pretty much the exact accuracy the original poster got, cool!

**Comment by mohits19** [comment]
Not sure about that error! We should look into that.
The good thing is it runs in reasonable amount of time. So we can test again if we are able to make any improvements to code!!

I am going to try and understand exactly what the code does. I vaguely understand it right now, but they may ask us questions about it!! 

Also, none of the code is our own. Is that ok? Should we try and adding/modifying something.

**Comment by mohits19** [comment]
@cleebp Can you try to run with more epochs to see if we get better accuracy. Maybe around 50-100 epochs since it is not taking that much time to run that!!

Also can you push muchdata.npy file so that we can also try running on our machines and VCL.

**Comment by cleebp** [comment]
Yeah I'll try running with more epoch's, im not sure if it will improve anything so i will start small (i dont really understand what they are too well). The original link you sent with the code goes through his code/process pretty well, I'm gonna start editing the latex document's methods section explaining it. I think the code is fine mostly not being our own but maybe we can comment it better and add stuff to the readme with clear instructions on how to run etc.

I will push code in a sec, trying out some of the edits people suggest in the comments to see if it improves anything.

**Comment by mohits19** [comment]
Awesome!!

Rishabh and me are trying to go through the code as well to understand it.
The code doesn't seem to do much preprocessing or any ROI selection. We were thinking by adding these things maybe we could make it better.
The link does seem to explain preprocessing but I can't seem to figure out how he has implemented it. Can you figure it out??

**Comment by cleebp** [comment]
Yeah they could help, from other similar approaches it seems like adding them doesnt tend to increase the accuracy by much, for the xgboost stuff for example people that added them got worse accuracies. All the same we could add them and compare the results.

The first comment on the notebook changed the chunks method in preprocessing so im adding that then will run again to get muchdata and push it all.

I'll look at preprocessing closer while its running.

**Comment by cleebp** [comment]
Also I'll be putting the new muchdata on the shared drive folder since itll be like 700mb...

**Comment by cleebp** [comment]
Fixing that chunk sizing bit got rid of my errors running cnn! Also I'm in the process of uploading muchdata to drive, will let yall know when its up.

**Comment by cleebp** [comment]
muchdata should be on the google drive now! Also I just pushed my changes to the branch. Additionally I just ran with 20 epochs and got an accuracy of .48 so I think 10 is better.

**Comment by mohits19** [comment]
Well done @cleebp . Let's stick with 10.
I will also try pulling your code and muchdata.npy and see if I can make some improvements. Otherwise, I think we still have enough to write about in the report.
I was thinking, in the Results and discussions, section we can do comparisons of accuracy vs no of epochs and show that 10 (or some other number) is optimal. 
Also we can try to show accuracy with and without preprocessing, ROI selection. (if we have time)
We have 2 days to write up all that which should be enough time.

**Comment by cleebp** [comment]
Yeah, I'm trying out a new way to run cnn right now where I keep running it until the desired accuracy is achieved - trying to get 0.69 accuracy but it can be easily modified to let it run to it finds any accuracy. If it ever does find 0.69 I will save that output. I'm gonna try to save each iteration's results that way we can get a table of outputs if it ends up taking awhile.

Also is everything for sure due wednesday?

**Comment by mohits19** [comment]
The posters are definitely due only next week!
The syllabus says report is due on 19th but no other mention of deadline anywhere else!!

**Comment by cleebp** [comment]
I'll make a post in the support forum and try to get clarification.

**Comment by cleebp** [comment]
Heres what the accuracies look like after 14 trials:

0.56
0.56
0.47
0.57
0.59
0.59
0.54
0.53
0.59
0.52
0.6
0.67
0.55
0.61

avg = 0.568

I think thats a pretty normal result, the xgboost submission was getting ~0.57 accuracy so our average makes sense with the data most likely.

**Comment by cleebp** [comment]
So i think the code is all in a good state, we could add some comments but for now its alright - I went ahead and merged the CNN work into master and cleaned up the orphaned files we werent using anymore and worked on the readme a bit. Going to close this thread now since the main priorities left are the report and poster.
